File: ncPersistencia/pPix.vb
Imports ncComum
Imports ncComum.nsAcessoBD
Imports ncComum.nsExcecao
Imports ncComum.nsFuncoes
Imports ncDados


Public Class ColecaoPix
    Inherits List(Of dPix)
End Class
Public Class pPix

    Public Function Consultar(ByVal dados As dPix) As ColecaoPix

        Dim retorno As ColecaoPix
        Dim acessoBanco As cAcessoBD
        Dim ds As DataSet
        Dim dt As DataTable
        Dim row As DataRow
        Dim item As dPix
        Dim sqlSelect As String
        Dim sqlWhere As String
        Dim sqlFrom As String

        Try

            acessoBanco = New cAcessoBD

            sqlSelect = " Select txID, Observacao, SolicitacaoPagador as pagador, Original, controle "

            sqlWhere = String.Empty
            sqlFrom = " From PIX "

            '-- TxId
            If dados.TxId <> 0 Then
                sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.TxId, "TxId")
            End If

            '-- TxId
            If dados.Controle <> 0 Then
                sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.Controle, "controle")
            End If


            If Not sqlWhere.Equals(String.Empty) Then
                sqlWhere = " WHERE " & sqlWhere
            End If

            ds = acessoBanco.ExecutarDS(sqlSelect & " " & sqlFrom & " " & sqlWhere)


            If Not ds Is Nothing Then
                If ds.Tables.Count > 0 Then
                    dt = ds.Tables(0)

                    If dt.Rows.Count > 0 Then
                        retorno = New ColecaoPix

                        For Each row In dt.Rows
                            item = New dPix
                            item.TxId = cFuncoes.RetornarTexto(row("TxId"))
                            'item.Cliente = cFuncoes.RetornarTexto(row("Cliente"))
                            'item.Cpf = cFuncoes.RetornarTexto(row("Cpf"))
                            'item.Cnpj = cFuncoes.RetornarTexto(row("Cnpj"))
                            item.Observacao = cFuncoes.RetornarTexto(row("Observacao"))
                            item.Pagador = cFuncoes.RetornarTexto(row("pagador"))
                            item.Original = cFuncoes.RetornarDecimal(row("Original"))
                            item.Controle = cFuncoes.RetornarInteiro(row("controle"))
                            retorno.Add(item)
                        Next
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If
            Else
                retorno = Nothing
            End If

        Catch ex As Exception

            retorno = Nothing
            Throw New ExcecaoNascomercio("Erro em Consultar Pix[" & Me.ToString() & "] - " & ex.Message)

        End Try

        Return retorno

    End Function


    Public Function Consultar() As dPix

        Dim retorno As dPix
        Dim acessoBanco As cAcessoBD
        Dim ds As DataSet
        Dim dt As DataTable
        Dim row As DataRow
        Dim item As dPix
        Dim sqlSelect As String
        Dim sqlWhere As String
        Dim sqlFrom As String

        Try

            acessoBanco = New cAcessoBD

            sqlSelect = " Select ID, txID, SolicitacaoPagador, Original, status, Observacao, DataHora, controle   "

            sqlWhere = " order by DataHora desc LIMIT 1;"
            sqlFrom = " From PIX "

            ds = acessoBanco.ExecutarDS(sqlSelect & " " & sqlFrom & " " & sqlWhere)


            If Not ds Is Nothing Then
                If ds.Tables.Count > 0 Then
                    dt = ds.Tables(0)

                    If dt.Rows.Count > 0 Then
                        retorno = New dPix

                        For Each row In dt.Rows
                            item = New dPix
                            item.ID = cFuncoes.RetornarTexto(row("ID"))
                            item.TxId = cFuncoes.RetornarTexto(row("TxId"))
                            item.Original = cFuncoes.RetornarDecimal(row("Original"))
                            item.Status = cFuncoes.RetornarTexto(row("status"))
                            item.DataHora = cFuncoes.RetornarData(row("DataHora"))
                            item.Observacao = cFuncoes.RetornarTexto(row("Observacao"))
                            item.Controle = cFuncoes.RetornarInteiro(row("controle"))
                            retorno = item
                        Next
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If
            Else
                retorno = Nothing
            End If

        Catch ex As Exception

            retorno = Nothing
            Throw New ExcecaoNascomercio("Erro em Consultar Pix[" & Me.ToString() & "] - " & ex.Message)

        End Try

        Return retorno

    End Function


    Public Function ConsultarConfig() As dPixConfig

        Dim retorno As dPixConfig
        Dim acessoBanco As cAcessoBD
        Dim ds As DataSet
        Dim dt As DataTable
        Dim row As DataRow
        Dim item As dPixConfig
        Dim sqlSelect As String
        Dim sqlWhere As String
        Dim sqlFrom As String

        Try

            acessoBanco = New cAcessoBD
            sqlSelect = " SELECT Cliente, Cpf, Cnpj, Nome, Chave "
            sqlWhere = String.Empty
            sqlFrom = "  FROM pixconfig "

            ds = acessoBanco.ExecutarDS(sqlSelect & " " & sqlFrom)

            If Not ds Is Nothing Then
                If ds.Tables.Count > 0 Then
                    dt = ds.Tables(0)

                    If dt.Rows.Count > 0 Then
                        retorno = New dPixConfig

                        For Each row In dt.Rows
                            item = New dPixConfig
                            item.Cliente = cFuncoes.RetornarTexto(row("Cliente"))
                            item.Cpf = cFuncoes.RetornarTexto(row("Cpf"))
                            item.Cnpj = cFuncoes.RetornarTexto(row("Cnpj"))
                            item.Nome = cFuncoes.RetornarTexto(row("Nome"))
                            item.Chave = cFuncoes.RetornarTexto(row("Chave"))
                        Next
                        retorno = item
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If
            Else
                retorno = Nothing
            End If

        Catch ex As Exception

            retorno = Nothing
            Throw New ExcecaoNascomercio("Erro em Consultar Pix[" & Me.ToString() & "] - " & ex.Message)

        End Try

        Return retorno

    End Function

    Public Function Incluir(ByVal dados As dPix) As Integer
        Dim colecaoPRODCOR As ColecaoPix = Nothing
        Dim retorno As Integer
        Dim acessoBanco As cAcessoBD
        Dim comandoSQL As String
        Dim ds As DataSet
        Dim dt As DataTable
        Dim ColecaoPix As List(Of String)
        Dim row As DataRow
        Dim item As String
        Dim ret As ColecaoPix

        Try

            acessoBanco = New cAcessoBD

            comandoSQL = " INSERT INTO PIX (ID, SolicitacaoPagador, Original, DataHora, Observacao, controle)  VALUES ("
            comandoSQL += cFuncoes.PersistirTexto(DateTime.Now.ToString("yyyyMMddHHmmss")) + "," + cFuncoes.PersistirTexto(dados.Pagador) & "," & cFuncoes.PersistirDecimal(dados.Original)
            comandoSQL += "," & cFuncoes.PersistirDataHora(DateTime.Now.ToString("dd/MM/yyyy HH:mm:ss")) & "," & cFuncoes.PersistirTexto(dados.Observacao) & "," & cFuncoes.PersistirInteiro(dados.Controle) + ")"

            retorno = acessoBanco.ExecutarCID(comandoSQL)

            comandoSQL = " SELECT controle FROM pix where ID = (select MAX(ID) from pix);"
            ds = acessoBanco.ExecutarDS(comandoSQL)

            If Not ds Is Nothing Then
                If ds.Tables.Count > 0 Then
                    dt = ds.Tables(0)

                    If dt.Rows.Count > 0 Then
                        ret = New ColecaoPix()

                        For Each row In dt.Rows
                            retorno = nsFuncoes.cFuncoes.RetornarTexto(row("controle"))
                        Next
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If
            Else
                retorno = Nothing
            End If

        Catch ex As Exception

            retorno = Nothing
            Throw New ExcecaoNascomercio("Erro em Incluir Pix [" & Me.ToString() & "] - " & ex.Message)

        End Try

        Incluir = retorno

    End Function

End Class

File: ncRepositorios/pPix.cs
using System;
using System.Collections.Generic;
using Dapper;
using MySql.Data.MySqlClient;
using ncNComum;
using ncModelos;
using ncRepositorios;

public class ColecaoPix : System.Collections.Generic.List<dPix> { }

public class pPix : RepositorioBase, IpPix
{
    public ColecaoPix Consultar(dPix dados)
    {
        try
        {
            using (var conn = CriarConexao())
            {
                var conditions = new List<string>();
                var p = new DynamicParameters();
                if (dados.TxId != "0") { conditions.Add("TxId=@TxId"); p.Add("TxId", dados.TxId); }
                if (dados.Controle != 0) { conditions.Add("controle=@Controle"); p.Add("Controle", dados.Controle); }
                var where = conditions.Count > 0 ? "WHERE " + string.Join(" AND ", conditions) : "";
                var sql = $"SELECT txID AS TxId, Observacao, SolicitacaoPagador AS Pagador, DataHora, Original, controle AS Controle, Status FROM PIX {where} ORDER BY DataHora DESC";
                var lista = conn.Query<dPix>(sql, p).AsList();
                if (lista.Count == 0) return null;
                var retorno = new ColecaoPix();
                retorno.AddRange(lista);
                return retorno;
            }
        }
        catch (ExcecaoNascomercio) { throw; }
        catch (Exception ex)
        {
            throw new ExcecaoNascomercio("Erro em Consultar Pix[" + ToString() + "] - " + ex.Message);
        }
    }

    public dPix Consultar(string tx)
    {
        try
        {
            using (var conn = CriarConexao())
            {
                var p = new DynamicParameters();
                var where = "";
                if (!string.IsNullOrEmpty(tx)) { where = "WHERE txID=@tx"; p.Add("tx", tx); }
                var sql = $"SELECT ID, txID AS TxId, SolicitacaoPagador AS Pagador, Original, status AS Status, Observacao, DataHora, controle AS Controle, UrlPix FROM PIX {where} ORDER BY DataHora DESC LIMIT 1";
                return conn.QueryFirstOrDefault<dPix>(sql, p);
            }
        }
        catch (ExcecaoNascomercio) { throw; }
        catch (Exception ex)
        {
            throw new ExcecaoNascomercio("Erro em Consultar Pix[" + ToString() + "] - " + ex.Message);
        }
    }

    public dPixConfig ConsultarConfig()
    {
        try
        {
            using (var conn = CriarConexao())
            {
                return conn.QueryFirstOrDefault<dPixConfig>(
                    "SELECT Banco, Cliente, Cpf, Cnpj, Nome, Chave, Client_id AS ClientID, client_secret AS ClientSecret, PathCertificate AS CertPath, PassCertificate AS CertPass, Email FROM pixconfig");
            }
        }
        catch (ExcecaoNascomercio) { throw; }
        catch (Exception ex)
        {
            throw new ExcecaoNascomercio("Erro em Consultar Pix[" + ToString() + "] - " + ex.Message);
        }
    }

    public int Incluir(dPix dados)
    {
        try
        {
            using (var conn = CriarConexao())
            {
                string id = DateTime.Now.ToString("yyyyMMddHHmmss");
                string datahora = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss");
                conn.Execute("INSERT INTO PIX (ID, SolicitacaoPagador, Original, DataHora, Observacao, controle) VALUES (@id, @Pagador, @Original, @datahora, @Observacao, @Controle)",
                    new { id, dados.Pagador, dados.Original, datahora, dados.Observacao, dados.Controle });
                return conn.QueryFirstOrDefault<int>("SELECT controle FROM pix WHERE datahora = (SELECT MAX(datahora) FROM pix)");
            }
        }
        catch (ExcecaoNascomercio) { throw; }
        catch (Exception ex)
        {
            throw new ExcecaoNascomercio("Erro em Incluir Pix [" + ToString() + "] - " + ex.Message);
        }
    }

    public int IncluirPixConfig(dPixConfig dados)
    {
        try
        {
            using (var conn = CriarConexao())
            {
                conn.Execute("INSERT INTO pixconfig (Banco, Cliente, Cpf, Cnpj, Nome, chave, Client_id, client_secret, PathCertificate, PassCertificate, Email) VALUES (@Banco, @Cliente, @Cpf, @Cnpj, @Nome, @Chave, @ClientID, @ClientSecret, @CertPath, @CertPass, @Email)",
                    new { dados.Banco, dados.Cliente, dados.Cpf, dados.Cnpj, dados.Nome, dados.Chave, dados.ClientID, dados.ClientSecret, dados.CertPath, dados.CertPass, dados.Email });
                return (int)conn.ExecuteScalar<long>("SELECT LAST_INSERT_ID()");
            }
        }
        catch (ExcecaoNascomercio) { throw; }
        catch (Exception ex)
        {
            throw new ExcecaoNascomercio("Erro em Incluir Pix [" + ToString() + "] - " + ex.Message);
        }
    }

    public int AlterarPixConfig(dPixConfig dados)
    {
        try
        {
            using (var conn = CriarConexao())
            {
                return conn.Execute("UPDATE pixconfig SET Banco=@Banco, Cpf=@Cpf, Cnpj=@Cnpj, Nome=@Nome, chave=@Chave, Client_id=@ClientID, client_secret=@ClientSecret, PathCertificate=@CertPath, PassCertificate=@CertPass",
                    new { dados.Banco, dados.Cpf, dados.Cnpj, dados.Nome, dados.Chave, dados.ClientID, dados.ClientSecret, dados.CertPath, dados.CertPass });
            }
        }
        catch (ExcecaoNascomercio) { throw; }
        catch (Exception ex)
        {
            throw new ExcecaoNascomercio("Erro em Alterar Configuração pix [" + ToString() + "] - " + ex.Message);
        }
    }
}
